Support for label capacities

frolv opened the pull request.

frolv committed:
Support for label capacities

Implemented an option through which a user can specify a capacity for a
label within an executable and have Bloaty report how much space they
are using relative to this capacity.

frolv committed:
Per-datasource label capacities

Changed label capacity definitions to be specific to data sources
instead of global.

keir commented on src/bloaty.cc:
```diff
@@ -291,6 +291,16 @@ std::string NameMunger::Munge(string_view name) const {
+      : name(name), vm_capacity(vm_capacity), file_capacity(file_capacity) {}
+
+  std::string name;
+  uint64_t vm_capacity = 0;
```
Please add mention of the units (bytes). If there is special meaning for some values (e.g.0) please document.

keir commented on src/bloaty.proto:
```diff
@@ -98,3 +101,12 @@ message Regex {
+// binaries in which there is a limit to how much space can be used.
+message LabelCapacity {
+  optional string label = 1;
+  optional uint64 vm_capacity = 2;
```
Document units (both).

keir commented on src/bloaty.proto:
```diff
@@ -98,3 +101,12 @@ message Regex {
+
+// Domain size capacities for a label in a data source.  When provided, the
+// output will display sizes as a percentage of this capacity.  Useful for
+// binaries in which there is a limit to how much space can be used.
```
Consider adding a comment showing an example; this can get as specific as the segment data source, with RAM and FLASH labels.

keir commented on src/bloaty.proto:
```diff
@@ -98,3 +101,12 @@ message Regex {
+// output will display sizes as a percentage of this capacity.  Useful for
+// binaries in which there is a limit to how much space can be used.
+message LabelCapacity {
+  optional string label = 1;
```
Perhaps mention in a comment that this must be an exact match.
Maybe drop a TODO to consider making this a regex.

frolv committed:
Only show limits for labels with capacity defined

Instead of printing the capacity of each label in the output, only print
capacities for the labels which explicitly have capacities set through a
config file.
